cpp/src/arrow/util/tracing_internal.cc

```diff
@@ -53,6 +60,43 @@ namespace {
 
 namespace sdktrace = opentelemetry::sdk::trace;
 
+// Custom JSON stdout exporter. Leverages the OTLP HTTP exporter's
+// utilities to log the same format that would be sent to OTLP.
```

Comment:
Is there a feature request on the OpenTelemetry side for them to expose this?

Reply:
There is https://github.com/open-telemetry/opentelemetry-cpp/pull/1111 which would effectively accomplish the same thing. However if they do add it to the contrib repo that would involve some more packaging work for us.

Reply:
The least riskiest thing would be to just define our own JSON format and not try to reuse any upstream components.

Reply:
If that's not too much work that would sound reasonable to me.

Reply:
Sounds good. CC @cpcloud any thoughts? There was such an exporter on the original PR (https://github.com/apache/arrow/pull/10260#discussion_r736884015) but as I recall the recommendation was to use the OTLP collector instead, so it was dropped. However, it seems the collector would be a lot more work for Conbench to integrate + is less convenient for local development workflows.

Reply:
Interesting. Typically, the collector usage is deploying a container alongside the application, so without knowing anything else I'm not sure why it would be a lot more work.

@lidavidm and I went back and forth on the original PR about this, but I'm generally -1 on adding an exporter because exporters force a choice when used inside of a library.

Libraries should *just* instrument ideally, and testing can be done by constructing exporters in tests.

Including an exporter inside a library can also easily conflict with an applications N other exporters.

If these exporters are included I think they should be restricted to only when tests are being compiled.

Reply:
Not totally understanding why a custom JSON format is less risky than using an upstream component or just having some documentation showing how to use the collector

Reply:
> Not totally understanding why a custom JSON format is less risky than using an upstream component or just having some documentation showing how to use the collector

Sorry, I meant just in terms of upstream changing APIs on us or something like that. The approach taken here does use the upstream generated Protobuf code and they might hide it from the public headers in the future, for instance.

> Libraries should just instrument ideally, and testing can be done by constructing exporters in tests.

That is the intent, but I think Arrow occupies a halfway point between library and application, especially when bindings come into play. You can't enable C++ exporters from Python or R, for instance. 

The exporters are not configured by default, for what it's worth - but they can be enabled by env var. The intent is that for development or testing, the env var can be used to easily dump spans somewhere, but an application would not use the env var and would configure its own exporter/tracer provider.

That said, this would all mostly be moot if Conbench ran a collector alongside test runs.

Reply:
> You can't enable C++ exporters from Python or R, for instance.

/me grumbles. I forgot about that. Ugh, I guess that means you'd probably have to use context propagation in-process which is kind of gross.

Reply:
Also see: https://github.com/open-telemetry/community/discussions/734

Yeah, in-process propagation is gross, the original PR used to have an example of that (to instrument Flight in Python) but I dropped it in the name of slimming the PR down.